shared/src/commonMain/kotlin/io/customer/shared/di/KMMComponent.kt

```diff
@@ -32,16 +45,61 @@ class KMMComponent(
     private val databaseHelper: DatabaseHelper
         get() = getSingletonInstance { DatabaseHelper(databaseDriverFactory = databaseDriverFactory) }
 
+    internal val jsonAdapter: JsonAdapter
+        get() = getSingletonInstance {
+            JsonAdapterImpl(
+                logger = staticComponent.logger,
+                sdkComponent = sdkComponent,
+            )
+        }
+
     internal val trackingTaskQueryHelper: TrackingTaskQueryHelper
         get() = getSingletonInstance {
             TrackingTaskQueryHelperImpl(
                 logger = staticComponent.logger,
                 dateTimeUtil = staticComponent.dateTimeUtil,
-                jsonAdapter = staticComponent.jsonAdapter,
+                jsonAdapter = jsonAdapter,
                 databaseUtil = staticComponent.databaseUtil,
                 workspace = sdkComponent.customerIOConfig.workspace,
                 backgroundQueueConfig = sdkComponent.customerIOConfig.backgroundQueue,
                 trackingTaskDAO = databaseHelper.trackingTaskDAO,
             )
         }
+
+    internal val trackingHttpClient: TrackingHttpClient
+        get() = getSingletonInstance {
+            TrackingHttpClientImpl(
+                logger = staticComponent.logger,
+                httpClient = httpClient,
+            )
+        }
+
+    private fun getHttpClientBuilder(): HttpClientBuilder = HttpClientBuilderImpl(
+        logger = staticComponent.logger,
+        sdkLogLevel = sdkComponent.customerIOConfig.sdkLogLevel,
+        workspace = sdkComponent.customerIOConfig.workspace,
+        networkConfig = sdkComponent.customerIOConfig.network,
+        userAgentStore = sdkComponent.userAgentStore,
```

Comment:
I am not ready to have an in-depth conversation about the implementation of our DI graph yet, but there is something here I wanted to mention because it's a great example. 

Do you remember in a previous PR when you asked me why I refactored the iOS SDK to using just 1 DI graph instead of multiple? 

This code demonstrates one of the reasons why I did. When there are multiple DI graphs, the implementation begins to get complex to read and write. For each dependency, you have to ask yourself, "What graph does this dependency exist? Where should I get it from?" as you see here in this highlighted snippet.  

I would rather see this code be simple to write and maintain like this:

```kotlin
        logger = logger,
        jsonAdapter = jsonAdapter,
        userAgentStore = userAgentStore
```

Especially when writing automated tests is where multiple graphs gets really complex. When you need to override a dependency in a graph with a mocked instance, you might have test code that looks like:

```kotlin
fun setup() {
  staticDiGraph.overrideDependency(...)
  diGraph.overrideDependency(...)
  otherDiGraph.overrideDependency(...)
}
```

When it is so much easier to write:

```kotlin
fun setup() {
  diGraph.overrideDependency(...)
  diGraph.overrideDependency(...)
  diGraph.overrideDependency(...)
}
````

I can understand that having every dependency listed in 1 source code file can be an eye sore. I can understand the value in having 1 DI graph for when the SDK is not initialized and when it is initialized. 

However, I do not see value in having multiple separate DI graphs when the SDK is initialized. I find it hard to read, maintain, engineering overkill, and debug. 

---

I am not sure if there is an action item from this PR comment of mine as I don't want this comment to be a blocker for this PR. However, I still wanted to bring it to your attention.

Reply:
Thanks for sharing this Levi. I do understand the concern. My plan currently is to have something like [koin scopes](https://insert-koin.io/docs/reference/koin-core/scopes) or [dagger subcomponents](https://dagger.dev/dev-guide/subcomponents.html) kind of concepts in our base DIGraph without any external dependency. But I'm not sure how much complexity that will add. But yes, this is something we need to work on and can achieve our goals without adding much complexity. I'm currently adding it to action items list, but once we start working on it, I think we can create a separate ticket for it, list our goals and challenges, share our proposals and finalize the implementation.

Reply:
Reading it again, I see that the idea behind static graph is slightly unclear here. I understand this is because we don't a complete running sample of KMM, so trying to explain here with some examples.

>  I can understand the value in having 1 DI graph for when the SDK is not initialized and when it is initialized.

This exactly is the idea behind these graphs:
- `KMMStaticComponent` -> Static graph that can be initialized anytime, whether or not the the SDK is initialized
- `KMMComponent` -> Can only be initialized with SDK initialization

The KMM project currently uses in a single graph which I see is the main cause of confusion here. Since we are not holding graphs in KMM yet, this may not be very clear. But the idea behind using in the SDK was something like:

```
val logger = CustomerIOShared.instance().diGraph.logger
```

Where `CustomerIOShared` is would be a class that does not rely on anything, and can be initialized anywhere without any prerequisites. And `diGraph` can be an instance of `KMMStaticComponent`

While the `KMMComponent` is much like our current SDK graph and will require SDK to be initialized before.

```
// Initialize the SDK
val backgroundQueue = CustomerIO.instance().diGraph.backgroundQueue
```

So, for cases where SDK is initialized at app start before any of our class is accessed, both `KMMStaticComponent` and `KMMComponent` will be initialized together (e.g. in our native apps of Android and iOS)

But, for cases where SDK is initialized later (e.g. ReactNative or Flutter apps), `KMMStaticComponent` will be initialized whenever accessed first and `KMMComponent` will be initialized later with SDK initialization and reuse the already initialized instance of `KMMStaticComponent`.  The [`KMMStaticComponent.attachSDKConfig`](https://github.com/customerio/mobile-shared/blob/develop/shared/src/commonMain/kotlin/io/customer/shared/di/KMMStaticComponent.kt#L40) can help us attach any configurations lazily.

---

I know KMM currently does not reflect the exact idea behind these multiple graphs. But just wanted to share the thoughts I had in mind when designing this, so when we start working on SDKs integration, we can may be try implementing this in similar way first?

And yes, no particular action item for now here. But this will help us wthen we start think on multiple DI graphs.

Reply:
I appreciate trying to explain your ideas behind the current implementation. Your comments were helpful. 

Again, not to block this PR, but while we are on the conversation I wanted to add the thoughts in my mind. We can address my thoughts in a future PR. 

I understand what koin scopes and dagger subcomponents are, but I do not see a reason why we need to apply this concept in our SDK project. I don't currently see our SDK needing scoped dependencies like a mobile app might need that. 

I also don't see the need to have a dependency graph of any kind until the SDK is initialized. That is because our SDK today is designed to be very useless until the SDK is initialized. I don't see a need to change that? Our SDK's features only work after an SDK is initialized, so what type of dependencies are even needed before SDK initialization? 

So, applying my knowledge of our SDKs today, I do not currently see a need for these DI features. Until we see a need to add these types of features to our SDK's dependency injection strategy, I find the implementation to be over engineering and complex for maintaining the code.

Reply:
I understand the point, but removing dependency from SDK initialization will be helpful in wrapper SDKs (e.g. ReactNative, Flutter, etc.). I'm not just talking about the `logger` here, but I think it can also help us in registering callbacks or adding/customizing module settings without relying on SDK initialization as in almost all non-native SDKs, the SDK initialization is delayed and it creates different challenges.

Once we create a ticket for this and start discussing this more actively, I'll try to share some code samples and use cases to explain why I think putting everything behind SDK initialization is not a very good idea, primarily, for non-native SDKs.